### PR TITLE
ENG-19277, catch all throwables on both snapshot write service and sy…

### PR DESCRIPTION
**Pull request by luyangco**
…nc service thread.

Snapshot write and sync service are two threads that responsible for writing snapshot to disk.
They cordinatate through a semaphore on how many bytes the write thread can write before synced.
If there is any exception uncaught in those two threads, it may lead to permit leaking, or even
worse, deadlock in the write thread. Unfortunately the snapshot write thread is one of the vital
thread to kepp snapshot subsystem running, its deadlock would make snapshot unusable.

Change-Id: I94c25a38bdebe8b032790f9fa0f35dede409eb7a

### CODE REVIEW

File: src/frontend/org/voltdb/DefaultSnapshotDataTarget.java [hunk 1]
```diff
@@ -521,6 +529,12 @@ public Object call() throws Exception {
                     SNAP_LOG.error("Error while attempting to write snapshot data to file " + m_file, e);
                     m_writeFailed = true;
                     throw e;
```
Comment (JadenW) [inline src/frontend/org/voltdb/DefaultSnapshotDataTarget.java]: nitpick: IOException would be double logged.

File: src/frontend/org/voltdb/DefaultSnapshotDataTarget.java [hunk 2]
```diff
@@ -299,10 +303,14 @@ public void run() {
                         if (!(e instanceof java.nio.channels.AsynchronousCloseException )) {
                             SNAP_LOG.error("Error syncing snapshot", e);
                         } else {
-                            SNAP_LOG.debug("Asynchronous close syncing snasphot data, presumably graceful", e);
+                            SNAP_LOG.info("Asynchronous close syncing snasphot data, presumably graceful", e);
                         }
+                    } catch (Throwable t) {
+                        SNAP_LOG.error("Unexpected error while fsyncing snapshot data to file " + m_file, t);
+                    } finally {
+                        final int bytesSinceLastSync = m_bytesWrittenSinceLastSync.getAndSet(0);
```
Comment (JadenW) [inline src/frontend/org/voltdb/DefaultSnapshotDataTarget.java]: Why need to defer get bytesSinceLastSync in the finally?

Comment (luyangco) [reply src/frontend/org/voltdb/DefaultSnapshotDataTarget.java]: Don't have a real reason, `m_bytesWrittenSinceLastSync` is not used before the actual sync so I just want to move it closer to permit release.

File: src/frontend/org/voltdb/DefaultSnapshotDataTarget.java (hunk 1 above)
Comment (luyangco) [reply src/frontend/org/voltdb/DefaultSnapshotDataTarget.java]: Could you elaborate? Why IOException is double logged?

Comment (JadenW) [reply src/frontend/org/voltdb/DefaultSnapshotDataTarget.java]: I think IOException would be caught by the first catch block and been re-throwed and caught up again in the second catch block, right?

Comment (luyangco) [reply src/frontend/org/voltdb/DefaultSnapshotDataTarget.java]: Good catch, I'll revise that in another pull request. https://github.com/VoltDB/voltdb/pull/6983

**Comment by luyangco** [comment]
Branch is renamed to ENG-19277-snapshotLivelock-backportv7.9, check pull request https://github.com/VoltDB/voltdb/pull/6986 for latest update.

**Pull request closed without merging**